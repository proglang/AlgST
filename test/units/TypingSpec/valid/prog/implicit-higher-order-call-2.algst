calc : ({Int} -> Int) -> Int

-- TODO: It would be great if the following would work as well, but this seems
-- quite hard with our custom function application checking rule.
--
-- check : Int
-- check = calc (implicit[Int] + 10)

check2 : Int
check2 =
  let inner : {Int} -> Int =
        implicit[Int] + 10
  in 
  calc inner

synth : Int
synth =
  let r = calc (implicit[Int] + 10) in
  r
