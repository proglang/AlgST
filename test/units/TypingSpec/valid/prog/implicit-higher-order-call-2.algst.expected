calc#M.2
|
`- SignatureDecl
   |
   `- type
      |
      `- Type.Arrow U
         |
         +- Type.Arrow ? U
         |  |
         |  +- Type.Type
         |  |  |
         |  |  `- TypeRef
         |  |     |
         |  |     `- Int
         |  |
         |  `- Type.Type
         |     |
         |     `- TypeRef
         |        |
         |        `- Int
         |
         `- Type.Type
            |
            `- TypeRef
               |
               `- Int

check2#M.0
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.Exp
         |
         `- TcExpX.ValueCase
            |
            +- Exp.Abs U «implicit-arg»#M.5
            |  |
            |  +- Type.Type
            |  |  |
            |  |  `- TypeRef
            |  |     |
            |  |     `- Int
            |  |
            |  `- Exp.App
            |     |
            |     +- Exp.App
            |     |  |
            |     |  +- Exp.Var (+)#Builtin.19
            |     |  |
            |     |  `- Exp.App
            |     |     |
            |     |     +- Exp.TypeApp
            |     |     |  |
            |     |     |  +- Exp.Var implicit#Builtin.32
            |     |     |  |
            |     |     |  `- Type.Type
            |     |     |     |
            |     |     |     `- TypeRef
            |     |     |        |
            |     |     |        `- Int
            |     |     |
            |     |     `- Exp.Var «implicit-arg»#M.5
            |     |
            |     `- Exp.Lit
            |        |
            |        `- Lit.Int 10
            |
            `- inner#M.3
               |
               `- Exp.App
                  |
                  +- Exp.Var calc#M.2
                  |
                  `- Exp.Abs U «implicit-arg»#M.6
                     |
                     +- Type.Type
                     |  |
                     |  `- TypeRef
                     |     |
                     |     `- Int
                     |
                     `- Exp.App
                        |
                        +- Exp.Var inner#M.3
                        |
                        `- Exp.Var «implicit-arg»#M.6

synth#M.1
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.Exp
         |
         `- TcExpX.ValueCase
            |
            +- Exp.App
            |  |
            |  +- Exp.Var calc#M.2
            |  |
            |  `- Exp.Abs U «implicit-arg»#M.7
            |     |
            |     +- Type.Type
            |     |  |
            |     |  `- TypeRef
            |     |     |
            |     |     `- Int
            |     |
            |     `- Exp.App
            |        |
            |        +- Exp.App
            |        |  |
            |        |  +- Exp.Var (+)#Builtin.19
            |        |  |
            |        |  `- Exp.App
            |        |     |
            |        |     +- Exp.TypeApp
            |        |     |  |
            |        |     |  +- Exp.Var implicit#Builtin.32
            |        |     |  |
            |        |     |  `- Type.Type
            |        |     |     |
            |        |     |     `- TypeRef
            |        |     |        |
            |        |     |        `- Int
            |        |     |
            |        |     `- Exp.Var «implicit-arg»#M.7
            |        |
            |        `- Exp.Lit
            |           |
            |           `- Lit.Int 10
            |
            `- r#M.4
               |
               `- Exp.Var r#M.4

